### PR TITLE
Generic HTTP/2 workers

### PR DESCRIPTION
**Pull request by rslota**
This PR introduces a generic HTTP/2 workers that deduplicate lots of code from  FCM and APNS workers. The FCM implementation is from #52 . 
Also I've hanged HTTP/2 client library. This is due to fact that `Kadabra` is kinda slow and also, fails to deliver all messages when there are lot of streams open. In my load testing `Kadabra` showed many flaws, like dropped massages, corrupted connections, etc., while `chatterbox` is almost flawless (it need one small change: joedevivo/chatterbox#106).  

This PR is more like "to discuss" and may require more work.

### CODE REVIEW

**Comment by hpopp** [comment]
Nice. I was thinking about this exact refactor the other day.

I still need to look things over but first things that come to mind:

We used `chatterbox` up until `hex.pm` banned all GitHub dependencies. I wrote `kadabra` as sort of a stop-gap measure until something better came along. I noticed someone finally posted `chatterbox` to hex, but we'll need to test. It looks like it has very few downloads.

If we do migrate back to `chatterbox` we also need to make sure heartbeat pings work. `chatterbox` doesn't keep connections alive on its own, and I remember having extreme difficulty getting it to send a ping manually.

**Comment by rslota** [comment]
Hi,

In regard to low download count: this is na erlang project and erlang world still either don't know about hex or simply don't use rebar3 that allow to do that. In my load testing I've tested several HTTP/2 client libraries and `chatterbox` is the only one that works stable and fast. 
If you really need ping feature I can look into it and make a PR to `chatterbox` that enables that.

**Comment by rslota** [comment]
`chatterbox` ping, here you go: joedevivo/chatterbox#107

**Comment by asgoel** [comment]
@hpopp couldn't you also get around the chatterbox issue by not including it as an explicit dependency but forcing people who want to use pigeon to add it separately as a dep/application? There are definitely other libraries that do that as well to circumvent the hex.pm issue

**Comment by hpopp** [comment]
That workaround still feels hackish and goes against my general philosophy for published packages.

I wouldn't be opposed to some sort of configuration option that lets you specify the http2 client used. I'm not ready to give up on kadabra yet, but if others want to use chatterbox it should be perfectly acceptable.

**Comment by tsloughter** [comment]
I published chatterbox to hex and happy to keep cutting releases when updates you need get accepted into chatterbox.

**Comment by hpopp** [comment]
Awesome! The only outstanding required feature is https://github.com/joedevivo/chatterbox/pull/107. APNS will drop long-running connections if there's no activity.

**Comment by tsloughter** [comment]
Ah yea, merging that now.

**Comment by tsloughter** [comment]
Merged and I published version 0.4.2 to hex.

**Comment by hpopp** [comment]
Wrote an initial implementation of client adapters at #67. From my testing so far both of the `kadabra` and `chatterbox` adapters work fine.

I'm starting to lean more toward making `chatterbox` the default adapter. I want to update `kadabra` to use `:gen_statem` and the `Registry`, but it will bump the minimum required elixir version to 1.4.

File: lib/pigeon/apns_worker.ex [hunk 1]
```diff
@@ -2,95 +2,42 @@ defmodule Pigeon.APNSWorker do
   @moduledoc """
     Handles all APNS request and response parsing over an HTTP2 connection.
   """
-  use GenServer
+  use Pigeon.GenericH2Worker, ping_interval: 600_000
```
Comment (asgoel) [inline lib/pigeon/apns_worker.ex]: @rslota any chance you can make this ping_interval configurable? some networks (such as AWS) require activity at a higher frequency so the ping has to be lower

**Comment by asgoel** [comment]
@rslota how stable is this code compared to master? wondering if i could switch a production app to using this vs kadabra

**Comment by hpopp** [comment]
Updated to `v0.12.0` which allows optional `chatterbox` configuration. Check the README for details

**Comment by hpopp** [comment]
Closing with the published changes in `v1.1`

**Pull request closed without merging**